### PR TITLE
Remember last abbreviations used in wrap with abbreviation.

### PR DESCRIPTION
**Pull request by gushuro**
When executing the Wrap with Abbreviation (resp. Individual Lines one) command, now the inputBox comes prefilled with the previous value entered, and cursor at the end.

Also, made a change so that we don't modify preview in wrap with abbreviation when the abbreviation typed is incomplete.

See https://github.com/Microsoft/vscode/issues/23169 and https://github.com/Microsoft/vscode/issues/40096

### CODE REVIEW

**Comment by gushuro** [comment]
On second thoughts, we might want to have the whole previous abbreviation selected when invoking the command again, so it's easier to write a whole new one (and users can still press right arrow to un-select it).

**Comment by ledniy** [comment]
Really missing this after switching from Atom, any plans to release it?

File: extensions/emmet/src/abbreviationActions.ts [hunk 1]
```diff
@@ -161,13 +164,26 @@ function doWrapping(individualLines: boolean, args: any) {
 		let extractedResults = helper.extractAbbreviationFromText(inputAbbreviation);
 		if (!extractedResults) {
 			return Promise.resolve(inPreview);
-		} else if (extractedResults.abbreviation !== inputAbbreviation) {
-			// Not clear what should we do in this case. Warn the user? How?
+		} else if (extractedResults.abbreviation !== inputAbbreviation.replace(/\|[a-zA-Z]+/g, '')) {
```
Comment (rzhao271) [inline extensions/emmet/src/abbreviationActions.ts:167]: Sanity check: is the regex taking out filters? Might be worth adding a comment about it

**Comment by rzhao271** [comment]
I know it's been almost three years, but this actually looks like a nice workaround in the meantime considering how:

- Using a quickpick to show history is unintuitive https://github.com/microsoft/vscode/pull/114170
- Implementing history into quickinput requires changing the API and would take longer than merging in this patch https://github.com/microsoft/vscode/issues/18232 

@gushuro wondering if you can update the PR?

File: extensions/emmet/src/abbreviationActions.ts [hunk 2]
```diff
@@ -161,13 +164,26 @@ function doWrapping(individualLines: boolean, args: any) {
...
+				});
+			}
+			// If the text is not a complete abbreviation yet, we don't modify the preview.
+			return Promise.resolve(inPreview);
 		}
 
 		let { abbreviation, filter } = extractedResults;
 		if (definitive) {
 			const revertPromise = inPreview ? revertPreview() : Promise.resolve();
+			if (individualLines) {
```
Comment (rzhao271) [inline extensions/emmet/src/abbreviationActions.ts:182]: You can just assume `individualLines = true` now, so there's no need for this check

File: extensions/emmet/src/abbreviationActions.ts [hunk 3]
```diff
@@ -201,7 +217,8 @@ function doWrapping(individualLines: boolean, args: any) {
 		}
 		return '';
 	}
-	const abbreviationPromise = (args && args['abbreviation']) ? Promise.resolve(args['abbreviation']) : vscode.window.showInputBox({ prompt: 'Enter Abbreviation', validateInput: inputChanged });
+	let value = individualLines ? previousAbbreviationIndividualLines : previousAbbreviation;
```
Comment (rzhao271) [inline extensions/emmet/src/abbreviationActions.ts:220]: You can just assume `individualLines = true` now, so there's no need for this check

File: extensions/emmet/src/abbreviationActions.ts [hunk 4]
```diff
@@ -16,6 +16,9 @@ const inlineElements = ['a', 'abbr', 'acronym', 'applet', 'b', 'basefont', 'bdo'
 	's', 'samp', 'select', 'small', 'span', 'strike', 'strong', 'sub', 'sup',
 	'textarea', 'tt', 'u', 'var'];
 
+let previousAbbreviation = '';
+let previousAbbreviationIndividualLines = '';
```
Comment (rzhao271) [inline extensions/emmet/src/abbreviationActions.ts:20]: You can just assume `individualLines = true` now, so I think just having `previousAbbreviation` is ok.

**Review by rzhao271 — CHANGES_REQUESTED** [review]
Since this also changes the UX, I'm wondering whether it can be put behind an extension settings flag, in case a user is already used to directly typing an expansion into the inputbox.

**Review by rzhao271 — CHANGES_REQUESTED** [review]
Slight GitHub glitch, marking as Request changes again

**Comment by rzhao271** [comment]
Closing for now.

**Pull request closed without merging**